DEVDOCS-2034 (#608)
bigcommerce/dev-docs: docs/api-docs/storefront/widgets/widgets-overview.md

<div class="otp" id="no-index">

### On this page

- [Regions](#regions)
- [Widget templates](#widget-templates)
- [Widgets](#widgets)
- [Placements](#placements)
- [Placements and widgets](#placements-and-widgets)
- [Widgets on the storefront](#widgets-on-the-storefront)
- [Definitions](#definitions)
- [Resources](#resources)

</div>

The Widgets API allows developers to programmatically associate content with regions on a BigCommerce storefront. The content can consist of HTML, CSS, and JavaScript, and the API supports configuration via [Handlebars](https://handlebarsjs.com/) variables. It can support many types of content such as YouTube Videos, image sliders, and chat apps.

Some benefits are:
* Inject modular, reusable blocks of content inside new and existing store pages
* Build tools that allow non-technical users to control content without editing theme files
* Target specific products, categories or brands with widgets

In this article, we are going to review all the components that make a widget. Then, we’ll use what we’ve learned to create a widget.

## Regions

> [Regions](https://developer.bigcommerce.com/api-reference/storefront/widgets-api/regions/getcontentregions) are specific locations in the Stencil theme files where you can place a widget. A region is added at the file level using the format {{{region name="..."}}}. A region can be named however you like, but it is best practice to give it a name that is descriptive of the location and function. A theme file can have as many regions as you want, with more than one widget assigned to the region and the [Placement](/api-docs/storefront/widgets/widgets-overview#widgets_placements) `sort_order` controlling how the widgets appear on the storefront.

Reviewer: Looks like the Placements link needs to be updated. 
```suggestion
[Regions](https://developer.bigcommerce.com/api-reference/storefront/widgets-api/regions/getcontentregions) are specific locations in the Stencil theme files where you can place a widget. A region is added at the file level using the format {{{region name="..."}}}. A region can be named however you like, but it is best practice to give it a name that is descriptive of the location and function. A theme file can have as many regions as you want, with more than one widget assigned to the region and the [Placements](https://developer.bigcommerce.com/api-docs/store-management/widgets/overview#placements) `sort_order` controlling how the widgets appear on the storefront.
```

![Content Region](//s3.amazonaws.com/user-content.stoplight.io/6012/1551970794664 "Content Region")

![Content Region](//s3.amazonaws.com/user-content.stoplight.io/6012/1551970767918 "Content Region")

Most themes in the BigCommerce marketplace come with predefined regions. It is best to use those first. By editing the theme and adding theme regions, updates will need to be managed manually.

## Widget templates

> [Widget templates](/api-reference/storefront/widgets-api/widget-template/getwidgettemplates) are Handlebars-enabled HTML templates which define the structure of the widget on a page. Widget templates can use the built in handlebar helpers such as {{if}} and {{each}}. Using the helpers allows for conditional logic to be built into the Widget templates. Widget templates can be reused to build multiple widgets.

Reviewer: ```suggestion
[Widget templates](/api-reference/storefront/widgets-api/widget-template/getwidgettemplates) are Handlebars-enabled HTML templates which define the structure of the widget on a page. Widget templates can use the built-in Handlebars helpers such as {{if}} and {{each}}. Using the helpers allows for conditional logic to be built into the Widget templates. Widget templates can be reused to build multiple widgets.
```

### Widget template examples

**Simple list**

> The simple list template creates a list where each item in the list can have a different color. This example uses the [each block helper](https://handlebarsjs.com/api-reference/helpers.html) in Handlebars to loop through each item in the list and display it. The text and color are determined by handlebars placeholders that are set when creating the [widget](/api-reference/storefront/widgets-api/widget/createwidget).

Reviewer: ```suggestion
The simple list template creates a list where each item in the list can have a different color. This example uses the [each block helper](https://handlebarsjs.com/api-reference/helpers.html) in Handlebars to loop through each item in the list and display it. The text and color are determined by Handlebars placeholders that are set when creating the [widget](/api-reference/storefront/widgets-api/widget/createwidget).
```

## Widgets

[Widgets](/api-reference/storefront/widgets-api/widget/getwidgets) are a unit of content that are placed on specific pages in a Stencil theme. Widgets are made of a widget configuration written in JSON and a widget template UUID. Widgets are rendered as part of the HTML on the storefront.

![Widget](//s3.amazonaws.com/user-content.stoplight.io/6012/1551971053085 "Storefront Widget")

### Widget configuration

> A Widget configuration is the JSON payload that defines the content to be rendered with a given widget template. In the previous simple list example, the widget template loops over a list_items array using `{{#each}}`.

Reviewer: ```suggestion
A Widget configuration is the JSON payload that defines the content to be rendered with a given widget template. In the previous simple list example, the widget template loops over a `list_items` array using `{{#each}}`.
```

<!-- theme:  -->

### Reusing widget configuration
> Keep in mind that when reusing the `widget_configuration` the array will have to be called `list_items`. Otherwise, the widget will be created on the frontend, but there will be no data, so nothing is rendered. This is because `list_items` was defined when the widget template was originally created. The widget configuration name can be anything that is set when the widget_template is created.

</div>
</div>
</div>

> **List items example: incorrect**

Reviewer 2: ```suggestion
**List items example: Incorrect**
```

The example above uses `list_items_two`, on line four, in the configuration which is not the same as `list_items`. If `list_items` has already been established during the initial configuration it must be set as `list_items` when being used again.

> **List items example: correct**

Reviewer 2: ```suggestion
**List items example: Correct**
```

## Placements

[Placements](/api-reference/storefront/widgets-api/placement/createplacement) determine the region where the widget is placed and in what order. The order of the placement is controlled by the `sort_order` when creating the placement.

> A placement must be created in order to use a Widget on the storefront.

Reviewer: ```suggestion
A placement must be created in order to use a widget on the storefront.
```

Placements can be used with `sort_order` and region to determine placement in a theme.

> ### Placements `entity_id`

Reviewer 2: ```suggestion
### Placements entity_id
```
